File: waypoints-api/example/src/main/java/de/md5lukas/examplej/ExamplePlugin.java
package de.md5lukas.examplej;

import de.md5lukas.waypoints.api.WaypointsAPI;
import de.md5lukas.waypoints.api.WaypointsPlayer;
import de.md5lukas.waypoints.pointers.PointerManager;
import net.kyori.adventure.text.Component;
import org.bukkit.command.Command;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.Player;
import org.bukkit.plugin.java.JavaPlugin;
import org.jetbrains.annotations.NotNull;

public class ExamplePlugin extends JavaPlugin {

  private WaypointsAPI api;
  private PointerManager pointers;

  @Override
  public void onEnable() {
    api = getServer().getServicesManager().load(WaypointsAPI.class);
    if (api == null) {
      getServer().getPluginManager().disablePlugin(this);
      return;
    }
    pointers = getServer().getServicesManager().load(PointerManager.class);

    api.getPublicWaypoints().getAllWaypointsCF().thenAccept(waypoints -> {
      for (var waypoint : waypoints) {
        System.out.printf(
            "Public waypoint %s is at position %s%n", waypoint.getName(), waypoint.getLocation());
      }
    });
  }

  @Override
  public boolean onCommand(
      @NotNull CommandSender sender,
      @NotNull Command command,
      @NotNull String label,
      @NotNull String[] args) {
    if (sender instanceof Player player) {
      api.getWaypointPlayerCF(player.getUniqueId())
          .thenCompose(WaypointsPlayer::getAllWaypointsCF)
          .thenAccept(waypoints -> {
            for (var waypoint : waypoints) {
              waypoint
                  .getFolderCF()
                  .thenAccept(folder -> sender.sendMessage(Component.text(String.format(
                      "You have the waypoint %s in the folder %s",
                      waypoint.getName(), folder != null ? folder.getName() : "none"))));
            }
          });
    }

    return true;
  }
}

File: waypoints-api/example/src/main/kotlin/de/md5lukas/examplek/ExamplePlugin.kt
package de.md5lukas.examplek

import de.md5lukas.waypoints.api.WaypointsAPI
import de.md5lukas.waypoints.api.WaypointsPointerManager
import de.md5lukas.waypoints.pointers.PointerManager
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch
import net.kyori.adventure.text.Component
import org.bukkit.command.Command
import org.bukkit.command.CommandSender
import org.bukkit.entity.Player
import org.bukkit.plugin.java.JavaPlugin
import java.util.*
import kotlin.coroutines.EmptyCoroutineContext

class ExamplePlugin : JavaPlugin() {

  private lateinit var api: WaypointsAPI
  private lateinit var waypointsPointerManager: WaypointsPointerManager
  private lateinit var pointers: PointerManager

  override fun onEnable() {
    val localApi = server.servicesManager.load(WaypointsAPI::class.java)
    if (localApi === null) {
      server.pluginManager.disablePlugin(this)
      return
    }
    this.api = localApi
    this.waypointsPointerManager = server.servicesManager.load(WaypointsPointerManager::class.java)!!
    this.pointers = server.servicesManager.load(PointerManager::class.java)!!

    CoroutineScope(EmptyCoroutineContext).launch {
      api.publicWaypoints.getAllWaypoints().forEach { waypoint ->
        println("Public waypoint ${waypoint.name} is at position ${waypoint.location}")
      }
    }
  }

  override fun onCommand(
      sender: CommandSender,
      command: Command,
      label: String,
      args: Array<String>,
  ): Boolean {
    if (sender is Player) {
      if (args.isEmpty()) {
        CoroutineScope(EmptyCoroutineContext).launch {
          api.getWaypointPlayer(sender.uniqueId).getAllWaypoints().forEach { waypoint ->
            sender.sendMessage(
              Component.text(
                "You have the waypoint ${waypoint.name} in the folder ${waypoint.getFolder()?.name ?: "none"}"))
          }
        }
      } else {
        CoroutineScope(EmptyCoroutineContext).launch {
          api.getWaypointByID(UUID.fromString(args[0]))?.let { waypoint ->
            waypointsPointerManager.enable(sender, waypoint)
          }
        }
      }
    }

    return true
  }
}
